### PR TITLE
Registration forms

### PR DESCRIPTION
**Pull request by stefanor**
This adds registration forms, as a separate registration mechanism.

Data is saved into KeyValue pairs.

This particular example form should probably be genericised (wafer shouldn't care about DebConf specifics).

The KV API clearly needs work. Updating contents is painful :(

### CODE REVIEW

File: wafer/tickets/templates/wafer.tickets/claim.html [hunk 1]
```diff
@@ -2,9 +2,9 @@
 {% load i18n %}
 {% load crispy_forms_tags %}
 {% block content %}
-{% if not WAFER_REGISTRATION_OPEN %}
-  {% trans "Registration is not currently open" %}
-{% elif not can_register %}
+{% if WAFER_REGISTRATION_OPEN != True %}
```
Comment (drnlm) [inline wafer/tickets/templates/wafer.tickets/claim.html]: If WAFER_REGISTRATION_OPEN is no longer going to be a simple boolean flag, we should rename it to something more descriptive. Something like WAFER_REGISTRATION_STAGE or WAFER_REGISTRATION_MODE, perhaps.

**Comment by drnlm** [comment]
What's the current approach for getting an overview of the pre-registration data?

docs?

**Comment by stefanor** [comment]
> What's the current approach for getting an overview of the pre-registration data?

The registration Form class.

**Comment by drnlm** [comment]
That just shows the user their own answers, right? There doesn't seem to be an obvious way to get an overview of everyone has pre-registered and their requests.

**Comment by stefanor** [comment]
> That just shows the user their own answers, right

Yep. I can add a CSV export. But really, we need to expose KV on the API.

File: wafer/settings.py [hunk 2]
```diff
@@ -259,6 +259,13 @@
 
 # Set this to False to disable registration
 WAFER_REGISTRATION_OPEN = True
+# Can be 'ticket' for Quicket tickets or 'form' for a classic form
+WAFER_REGISTRATION_MODE = 'ticket'
+
+# For REGISTRATION_MODE == 'form': The form to present, and the Group to
+# associate the responses with in the KV store
+WAFER_REGISTRATION_FORM = 'wafer.users.forms.ExampleRegistrationForm'
+WAFER_REGISTRATION_GROUP = 'Registration'
```
Comment (drnlm) [inline wafer/settings.py]: We currently have management/commands/wafer_add_default_groups.py to easily add the page edit and talk review groups. The "Registration" group should arguably be added to that list, so it's in place out of the box.

File: wafer/users/forms.py [hunk 3]
```diff
@@ -43,3 +46,37 @@ def __init__(self, *args, **kwargs):
...
         exclude = ('user', 'kv')
+
+
+def get_registration_form_class():
+    return import_string(settings.WAFER_REGISTRATION_FORM)
+
+
+class ExampleRegistrationForm(forms.Form):
+    debcamp = fields.BooleanField(
+        label='Plan to attend DebCamp', required=False)
```
Comment (drnlm) [inline wafer/users/forms.py]: Should these be marked as translatable?

File: wafer/users/templates/wafer.users/registration.html [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{% extends "wafer/base.html" %}
+{% load i18n %}
+{% load crispy_forms_tags %}
+{% block content %}
+<h1>{% trans 'Registration:' %}</h1>
+{% crispy form %}
+{% endblock %}
```
Comment (drnlm) [inline wafer/users/templates/wafer.users/registration.html]: If the <user>/register url is going to be unconditionally accessible, this should check the status of WAFER_REGISTRATION_MODE and report something sensible for the 'ticket' case.

File: wafer/settings.py (hunk 2 above)
Comment (stefanor) [inline wafer/settings.py]: In that case I can drop this configuration option.

File: wafer/users/forms.py [hunk 5]
```diff
@@ -43,3 +46,37 @@ def __init__(self, *args, **kwargs):
...
+    debconf = fields.BooleanField(
+        label=_('Plan to attend DebConf'), required=False)
+    require_sponsorship = fields.BooleanField(
+        label=_('Will require sponsorship'), required=False)
+
+    def __init__(self, *args, **kwargs):
+        super(ExampleRegistrationForm, self).__init__(*args, **kwargs)
+        self.helper = FormHelper()
+        self.helper.layout = Layout(
+            Fieldset('Pre-Registration',
```
Comment (drnlm) [inline wafer/users/forms.py]: "Pre-Registration" should probably also be marked translatable.

**Comment by drnlm** [comment]
I'd still like to seem some docs added about this, but we can probably land this and fix things up later

**Comment by hodgestar** [comment]
:+1:

**Pull request merged**